packages/suite/src/views/onboarding/steps/Firmware/index.tsx

```diff
@@ -1,241 +1,107 @@
-    const [savedModel, setSavedModel] = useState(2);
-    const [imageLoaded, setImageLoaded] = useState(false);
-
-    useEffect(() => {
-        if (!device?.features?.major_version) return;
-        setSavedModel(device?.features?.major_version);
-    }, [device]);
-
-    const { status, btcOnly } = firmware;
-    const isConnected = !!device;
-    const isInBootloader = Boolean(device && device.features && device.mode === 'bootloader');
-    const btcOnlyAvailable = device?.features && device.firmwareRelease.release.url_bitcoinonly;
+    const getComponent = () => {
+        // edge case 1 - error
+        if (firmware.error) {
+            return {
+                Body: <ErrorStep.Body />,
+                BottomBar: <RetryButton onClick={() => resetReducer()} />,
+            };
+        }
+
+        // // edge case 2 - user has reconnected device that is already up to date
+        if (firmware.status !== 'done' && device?.firmware === 'valid') {
+            return {
+                Body: <NoNewFirmware.Body />,
+                BottomBar: <ContinueButton onClick={() => goToNextStep()} />,
+            };
+        }
+
+        switch (firmware.status) {
+            case 'initial':
+                return {
+                    Body: <OnboardingInitialStep.Body />,
+                    BottomBar: <OnboardingInitialStep.BottomBar />,
+                };
+            case 'check-seed':
+                return {
+                    Body: <CheckSeedStep.Body />,
+                    BottomBar: <CheckSeedStep.BottomBar />,
+                };
+            case 'waiting-for-bootloader':
+                return {
+                    Body: <ReconnectInBootloaderStep.Body />,
+                    BottomBar: <ReconnectInBootloaderStep.BottomBar />,
+                };
+            case 'waiting-for-confirmation':
+            case 'installing':
+            case 'started':
+            case 'downloading':
+            case 'check-fingerprint':
+            case 'wait-for-reboot':
+            case 'unplug':
+                return {
+                    Body: <FirmwareProgressStep.Body />,
+                    BottomBar: null,
+                };
+            case 'reconnect-in-normal':
+                return {
+                    Body: <ReconnectInNormalStep.Body />,
+                    BottomBar: <ReconnectInNormalStep.BottomBar />,
+                };
+            case 'partially-done':
+                return {
+                    Body: <PartiallyDoneStep.Body />,
+                    BottomBar: <ContinueButton onClick={resetReducer} />,
+                };
+            case 'done':
+                return {
+                    Body: <DoneStep.Body />,
+                    BottomBar: <ContinueButton onClick={() => goToNextStep()} />,
+                };
+
+            default:
+                // 'ensure' type completeness
+                throw new Error('');
```

Comment:
We shouldn't throw empty message, makes it harder to identify what went wrong when it happens.